src/masternodes/mn_checks.cpp

```diff
@@ -1029,6 +1079,406 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
+        }
+
+        submitdfchtlc.creationTx = tx.GetHash();
+        submitdfchtlc.creationHeight = height;
+        submitdfchtlc.expireHeight = height + submitdfchtlc.timeout;
+        
+        return mnview.ICXSubmitDFCHTLC(submitdfchtlc);
+    }
+
+    Res operator()(const CICXSubmitEXTHTLCMessage& obj) const {
+        auto res = CheckICXTx();
+        if (!res) {
+            return res;
+        }
+
+        CICXSubmitEXTHTLCImplemetation submitexthtlc;
+        static_cast<CICXSubmitEXTHTLC&>(submitexthtlc) = obj;
+
+        auto offer=mnview.GetICXMakeOfferByCreationTx(submitexthtlc.offerTx);
+        if (!offer) {
+            return Res::Err("order with creation tx %s does not exists!", submitexthtlc.offerTx.GetHex());
+        }
+
+        auto order=mnview.GetICXOrderByCreationTx(offer->orderTx);
+        if (!order) {
+            return Res::Err("order with creation tx %s does not exists!", offer->orderTx.GetHex());
+        }
+
+        if (submitexthtlc.htlcscriptAddress.empty())
+            return Res::Err("Invalid htlcscriptAddress, htlcscriptAddress is empty and it must be set!");
+        if (submitexthtlc.hash.IsNull())
+            return Res::Err("Invalid hash, htlc hash is empty and it must be set!");
+        if (!submitexthtlc.ownerPubkey.IsFullyValid())
+            return Res::Err("Invalid refundPubkey is not a valid pubkey!");
+        if (submitexthtlc.timeout == 0)
+            return Res::Err("Invalid timout, must be greater than 0!");
+        
+        if (order->orderType==CICXOrder::TYPE_INTERNAL)
+        {
+            if (submitexthtlc.amount != offer->amount)
+                return Res::Err("amount in ext htlc must match the amount in the offer - %s != %s!", ValueFromAmount(submitexthtlc.amount).getValStr(), ValueFromAmount(offer->amount).getValStr());
+            
+            bool found=false;
+            mnview.ForEachICXSubmitEXTHTLC([&found,&submitexthtlc](CICXOrderView::StatusTxidKey const & key, uint8_t i) {
+                if (key.first.second != submitexthtlc.offerTx)
+                    return false;
+                found=true;
+                return false;
+            },{CICXSubmitDFCHTLC::STATUS_OPEN,submitexthtlc.offerTx});
+            if (found)
+                return Res::Err("ext htlc already submitted!");
+
+            std::unique_ptr<CICXSubmitDFCHTLCImplemetation> dfchtlc;
+            mnview.ForEachICXSubmitDFCHTLC([&](CICXOrderView::StatusTxidKey const & key, uint8_t) {
+                if (key.first.second == submitexthtlc.offerTx)
+                {
+                    dfchtlc = mnview.GetICXSubmitDFCHTLCByCreationTx(key.second);
+                    return (false);
+                }
+                return false;
+            },{CICXSubmitDFCHTLC::STATUS_OPEN,submitexthtlc.offerTx});
+            if (!dfchtlc)
+                return Res::Err("offer %s needs to have dfc htlc submitted first, but no dfc htlc found!", submitexthtlc.offerTx.GetHex());
```

Comment:
This error message is in both branches, is the flow for `TYPE_INTERNAL` supposed to be different?

Reply:
For `TYPE_INTERNAL` orders (DFC/BTC) you must first submit DFC HTLC, and for `TYPE_EXTERNAL` you must first submit EXT HTLC.